neps/nep-0413.md

```diff
@@ -0,0 +1,142 @@
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Wallets must implement a `signMessage` method, which takes a `message` destined to a specific `receiver` and transform it into a verifiable signature.
+
+### Input Interface
+`signMessage` must implement the following input interface:
+
+```jsx
+interface SignMessageParams {
+  message: string ; // The message that wants to be transmitted.
+  receiver: string; // The receiver to whom the message is destined (e.g. "alice.near" or "myapp.com").
+  nonce: [u8; 32] ; // A nonce that uniquely identifies this instance of the message, denoted as a 32 bytes array (a fixed `Buffer` in JS/TS).
+  callbackUrl?: string; // Optional, applicable to browser wallets (e.g. MyNearWallet). The URL to call after the signing process. Defaults to `window.location.href`.
+}
+```
+
+### Structure
+`signMessage` must embed the input `message`, `receiver` and `nonce` into the following predefined structure:
+
+```rust
+struct Payload {
+  message: string; // The same message passed in `SignMessageParams.message`
+  nonce: [u8; 32]; // The same nonce passed in `SignMessageParams.nonce`
+  receiver: string; // The same receiver passed in `SignMessageParams.receiver`
+}
+```
+
+### Signature
+In order to create a signature, `signMessage` must:
+1. Convert the `Payload` into its [JSON JCS](https://www.rfc-editor.org/rfc/rfc8785) string representation (i.e. a JSON string, with alphabetically ordered attributes).
+2. Prepend the `NEP0413:` string to the result from step 1.
+3. Compute the `SHA256` hash of the result from step 2.
+4. Sign the resulting `SHA256` hash from step 3 using a **full-access** key.
+
+> If the wallet does not hold any `full-access` keys, then it must return an error.
+
+### Example
+Assuming that the `signMessage` method was invoked, and that:
+- The input `message` is `"hi"`
+- The input `nonce` is `[0,...,31]`
+- The input `receiver` is `"myapp.com"`
+- The wallet stores a full-access private key
+
+The wallet must construct and sign the following `SHA256` hash:
+
+```jsx
+sha256.hash(`NEP0413:` + `{"message":"hi","nonce":[0,...,31],"receiver":"myapp.com"}`)
+```
+
+### Output Interface
+`signMessage` must return an object containing the **base64** representation of the `signature`, and all the data necessary to verify such signature. 
+
+```jsx
+interface SignedMessage {
+  accountId: string; // The account name to which the publicKey corresponds as plain text (e.g. "alice.near")
+  publicKey: string; // The public counterpart of the key used to sign, expressed as a string with format "<key-type>:<base-64-key-bytes>"
+  signature: string; // The base64 representation of the signature.
```

Comment:
I believe it is just a typo here regarding base64 encoding for publicKey and signature as they are commonly base58-encoded in NEAR Protocol. I assume it is a typo as the prior discussion (https://github.com/near/NEPs/pull/413#discussion_r1021799176) referenced base58 examples.

NOTE: In NEAR Protocol, base58 encoding is commonly used to display and sometimes store public/private keys, signatures, and hashes. And even though there was an old discussion about migrating away from it (https://github.com/near/nearcore/issues/2100), there was no action taken. Yet, base64 is used for arbitrary-length values (e.g. function args, return values, etc).

Reply:
Interesting, this might be my misunderstanding actually, when looking at this key:

```ed25519:6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y```

the second part (i.e. `6TupyNrcHGTt5XRLmHTc2KGaiSbjhQi1KHtCXTgbcr4Y`) is encoded using base58?

Reply:
Yes, that is correct, that value is base58-encoded.

[Here is the reference to nearcore implementation](https://github.com/near/nearcore/blob/db094eee0f325d1c28b5da342e0bbc7964290544/core/crypto/src/signature.rs#L180-L188) where you can see that the `key_data` (32 bytes) are Bs58-encoded.